notes/overriding.md

```diff
@@ -0,0 +1,676 @@
+in the NixOS module system, it is possible to address this issue using custom
+merge functions. This is a bit less powerful, but in a good way: it forces the
+merge strategy to be uniform along each field, while overlays can do pretty much
+what they want.
+
+## Proposal
+
+We basically propose to adopt the same kind of merge, priority, and custom merge
+function-based scheme as the NixOS module system. One big difference with the
+NixOS module system is that it would be built-in in the language, being usable
+for any native record whatsoever, and having good interaction with the rest of
+the language features. The implementation would have first class access to
+locations, to the AST, the memory layout, the metadata, and so on.
+
+We first describe ideas at a high-level. Issues, details and challenges are
+discussed further in a dedicated section.
+
+### Recursive records & merging
+
+**This section defines the semantics, rather than an actual efficient
+implementation**.
+
+As before, it is useful to see recursive records represented as functions -- or
+constructors -- that take a `self` parameter and return a final non recursive record,
+in the same way as the original overriding mechanism of Nixpkgs (the non-existing
+syntax `def := value` is used to insist on the fact that we are defining new objects):
+
+```
+r = {
+  a = 1;
+  b = a + 1;
+}
+// Definition of the representation of r
+repr(r) := fun self => {
+  a = 1;
+  b = self.a + 1;
+}
+```
+
+Field access amounts to compute a fix-point:
+
+```
+repr(r).foo := let fix = repr(r) fix in fix.foo
+```
+
+Merging simply merges the underlying representations:
+
+```
+repr(r1) & repr(r2) := fun self => r1 self & r2 self
+```
+
+This brings the problem of scoping: should a record be able to access a yet
+undefined field because it is expected to be provided by a subsequent merge?
+Two possible approaches:
+
+ - **dynamic scoping**: records can reference fields that are not explicitly
+   defined locally, such as:
+   ```nickel
+   {a = b} & {b = 1}
+   ```
+   However, dynamic scoping have a number of issues, and is usually considered
+   bad practice.
+ - **lexical scoping**: as currently, require self-referenced fields to be
+   defined locally. Note that thanks to contracts, one can require the existence
+   of a field without defining it. For example, we could write the previous
+   example as:
+   ```nickel
+   {a = b, b | Num} & {b = 1}
+   ```
+   This is also a better practice to explicitly state the fields whose presence
+   is required.
+
+### Priorities
+
+Merge is fundamentally commutative. The problem is, not all fields should be
+treated the same: some are default values that ought to be overridden, some are
+high-priority values that ought to override. The solution is to use a priority
+system, encoding the precedence of each value while retaining commutativity
+(only the priority annotation counts, not the order in which values are merged).
```

Comment:
You want to be much more precise here. You really want to spell out the semantics of priorities. So basically the entire merge algorithm with priorities (you don't need to specify the set priorities for that, you just need that the set of priorities has an order (even a partial order will do)).

The question of how to merge records with different priorities may arise. If so it will then be a natural question about whether we want to just take the record with the highest priority, or take some more complex merge, which may remove the need for the leafy default thing.